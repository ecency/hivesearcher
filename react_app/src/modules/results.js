import {Record, Map, OrderedMap} from 'immutable';
import parseQuery from '../utils/parse-query';
import resultGroupKey from '../utils/result-group-key.js'
import axios from 'axios';
import {API_URL} from '../config';
import {SORT_CHOICES, PAGE_SIZE} from '../constants';

export const POSTS_FETCH_BEGIN = 'results/FETCH_BEGIN';
export const POSTS_FETCH_OK = 'results/FETCH_OK';
export const POSTS_FETCH_ERROR = 'result/POSTS_FETCH_ERROR';
export const POSTS_INVALIDATE = 'result/POSTS_INVALIDATE';


export const ResultGroupRecord = Record({
    hits: null,
    took: null,
    entries: Map({
        popularity: Map({
            scrollId: null,
            hasMore: true,
            loading: true,
            list: OrderedMap({})
        }),
        relevance: Map({
            scrollId: null,
            hasMore: true,
            loading: true,
            list: OrderedMap({})
        }),
        newest: Map({
            scrollId: null,
            hasMore: true,
            loading: true,
            list: OrderedMap({})
        })
    })
});


const initialState = Map();

export default (state = initialState, action) => {
    switch (action.type) {
        case POSTS_FETCH_BEGIN: {
            const {payload} = action;
            const {groupKey, sort} = payload;

            if (state.get(groupKey) === undefined) {
                return state.set(groupKey, new ResultGroupRecord());
            }

            return state.setIn([groupKey, 'entries', sort, 'loading'], true);
        }

        case POSTS_FETCH_OK: {
            const {groupKey, sort, data} = action.payload;
            const {results, hits, took, scroll_id: scrollId} = data;

            const realHits = state.getIn([groupKey, 'hits']) || hits;
            const realTook = state.getIn([groupKey, 'took']) || took;

            let newState = state
                .setIn([groupKey, 'hits'], realHits)
                .setIn([groupKey, 'took'], realTook)
                .setIn([groupKey, 'entries', sort, 'loading'], false)
                .setIn([groupKey, 'entries', sort, 'scrollId'], scrollId ? scrollId : null)
                .setIn([groupKey, 'entries', sort, 'hasMore'], (results.length === PAGE_SIZE));

            results.forEach(entry => {
                if (!newState.hasIn([groupKey, 'entries', sort, 'list', `${entry.id}`])) {
                    newState = newState.setIn(
                        [groupKey, 'entries', sort, 'list', `${entry.id}`],
                        entry
                    );
                }
            });

            return newState;
        }
        default:
            return state;
    }
}

export const fetchResults = (query, sort, fetchMore) => {
    return (dispatch, getState) => {

        const groupKey = resultGroupKey(query);

        const {results} = getState();


        if (!fetchMore && results.get(groupKey) && results.getIn([groupKey, 'entries', sort, 'list']).size > 0) {
            return;
        }

        dispatch({
            type: POSTS_FETCH_BEGIN,
            payload: {groupKey, sort}
        });

        const formData = new FormData();
        formData.set('q', query);
        formData.set('sort', sort);

        axios({
            method: 'post',
            url: `${API_URL}api/search`,
            data: formData,
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
                'Accept': 'application/json'
            }
        }).then(function (resp) {

            dispatch({
                type: POSTS_FETCH_OK,
                payload: {groupKey, sort, data: resp.data}
            });

        }).catch(function (resp) {

        }).then(() => {

        })
    }
};


export const invalidateResults = () => {
    return dispatch => {
        dispatch({
            type: POSTS_INVALIDATE,
            payload: {}
        });
    }
};