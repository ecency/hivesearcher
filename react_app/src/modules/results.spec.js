/* eslint-disable */
import {Map, Record} from "immutable";

import {fetchBegin, fetchOk, fetchError, invalidate} from './results'
import results from './results';

describe('Results', () => {
    it('(1) New group should be created after POSTS_FETCH_BEGIN', () => {

        let query = 'esteem type:post author:good-karma';
        let sort = 'popularity';
        let page = 1;

        let state = Map();

        let new_state = results(state, fetchBegin(query, sort, page));

        expect(new_state).toMatchSnapshot();
    });


    it('(2) Results should be loaded after POSTS_FETCH_OK', () => {

        let query = 'esteem type:post author:good-karma';
        let sort = 'popularity';
        let page = 1;

        let state = Map();

        let new_state = results(state, fetchBegin(query, sort, page));

        let data = {
            "hits": 2,
            "took": "0.012",
            "pages": 1,
            "results": [{"id": 1, "title": "foo"}, {"id": 2, "title": "bar"}]
        };

        let new_state2 = results(new_state, fetchOk(query, sort, page, data));

        expect(new_state2).toMatchSnapshot();
    });

    it('(3) err should be true after POSTS_FETCH_ERROR', () => {

        let query = 'esteem type:post author:good-karma';
        let sort = 'popularity';
        let page = 1;

        let state = Map();

        let new_state = results(state, fetchBegin(query, sort, page));

        let data = {
            "hits": 2,
            "took": "0.012",
            "pages": 1,
            "results": [{"id": 1, "title": "foo"}, {"id": 2, "title": "bar"}]
        };

        let new_state2 = results(new_state, fetchOk(query, sort, page, data));

        page = 2;

        let new_state3 = results(new_state2, fetchError(query, sort, page));

        expect(new_state3).toMatchSnapshot();
    });

    it('(4) err should be false after POSTS_FETCH_BEGIN', () => {

        let query = 'esteem type:post author:good-karma';
        let sort = 'popularity';
        let page = 1;

        let state = Map();

        let new_state = results(state, fetchBegin(query, sort, page));

        let data = {
            "hits": 2,
            "took": "0.012",
            "pages": 1,
            "results": [{"id": 1, "title": "foo"}, {"id": 2, "title": "bar"}]
        };

        let new_state2 = results(new_state, fetchOk(query, sort, page, data));

        page = 2;

        let new_state3 = results(new_state2, fetchError(query, sort, page));

        let new_state4 = results(new_state3, fetchBegin(query, sort, page));

        expect(new_state4).toMatchSnapshot();
    });


    it('(5) Group should be deleted after POSTS_INVALIDATE', () => {

        let query = 'esteem type:post author:good-karma';
        let sort = 'popularity';
        let page = 1;

        let state = Map();

        let new_state = results(state, fetchBegin(query, sort, page));

        query = 'esteem type:post author:good-karma';

        let new_state2 = results(new_state, invalidate(query));

        expect(new_state2).toMatchSnapshot();
    });

});