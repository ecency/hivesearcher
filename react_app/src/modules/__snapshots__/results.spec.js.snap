// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results (1) New group should be created after POSTS_FETCH_BEGIN 1`] = `
Immutable.Map {
  esteem type:post author:good-karma: Object {
    "err": false,
    "hits": null,
    "pages": 0,
    "scopes": Object {
      "popularity-1": Object {
        "list": Array [],
        "loading": true,
      },
    },
    "took": null,
  },
}
`;

exports[`Results (2) Results should be loaded after POSTS_FETCH_OK 1`] = `
Immutable.Map {
  esteem type:post author:good-karma: Object {
    "err": false,
    "hits": 2,
    "pages": 1,
    "scopes": Object {
      "popularity-1": Object {
        "list": Array [
          Object {
            "id": 1,
            "title": "foo",
          },
          Object {
            "id": 2,
            "title": "bar",
          },
        ],
        "loading": false,
      },
    },
    "took": "0.012",
  },
}
`;

exports[`Results (3) err should be true after POSTS_FETCH_ERROR 1`] = `
Immutable.Map {
  esteem type:post author:good-karma: Object {
    "err": true,
    "hits": 2,
    "pages": 1,
    "scopes": Object {
      "popularity-1": Object {
        "list": Array [
          Object {
            "id": 1,
            "title": "foo",
          },
          Object {
            "id": 2,
            "title": "bar",
          },
        ],
        "loading": false,
      },
      "popularity-2": Object {
        "loading": false,
      },
    },
    "took": "0.012",
  },
}
`;

exports[`Results (4) err should be false after POSTS_FETCH_BEGIN 1`] = `
Immutable.Map {
  esteem type:post author:good-karma: Object {
    "err": false,
    "hits": 2,
    "pages": 1,
    "scopes": Object {
      "popularity-1": Object {
        "list": Array [
          Object {
            "id": 1,
            "title": "foo",
          },
          Object {
            "id": 2,
            "title": "bar",
          },
        ],
        "loading": false,
      },
      "popularity-2": Object {
        "list": Array [],
        "loading": true,
      },
    },
    "took": "0.012",
  },
}
`;

exports[`Results (5) Group should be deleted after POSTS_INVALIDATE 1`] = `
Immutable.Map {
}
`;
